### ISSUE TITLE
No rest frame of cell frequency shift in calculation of bf opacity

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#73]
In radiation.c we have to calculate the bf opacity so we can work out a) the contribution from bf heating and b) how much to reduce the photon weight by over the distance travelled ds. 

This is done by cycling through the cross sections and checking if the photon frequency is above the photoionization edge. If so, then that cross section contributes to the opacity.

This frequency check is done incorrectly, as it compares the absolute frequency of the edge to the frequency of the photon, instead of shifting the edge into the rest frame of the cell. 

I will have to check if similar things are being done elsewhere.

the offending piece of code is here. One can see that we jsut compare the frequency of the photon to that of the edge with no velocity dependence.

``` C
if (freq > phot_freq_min)

    {

      if (geo.ioniz_or_extract) // 57h -- ksl -- 060715
    {           // Initialize during ionization cycles only
      for (nion = 0; nion < nions; nion++)
        {
          kappa_ion[nion] = 0;
          frac_ion[nion] = 0;
        }
    }
      /* Next section is for photoionization with Topbase.  There may be more
         than one x-section associated with an ion, and so one has to keep track
         of the energy that goes into heating electrons carefully.  */

      /* Next steps are a way to avoid the loop over photoionization x sections when it should not matter */
      if (DENSITY_PHOT_MIN > 0) // 57h -- ksl -- 060715
    {           // Initialize during ionization cycles only


      /* 57h -- 06jul -- ksl -- change loop to use pointers ordered by frequency */
      for (n = 0; n < ntop_phot; n++)
        {
          x_top_ptr = phot_top_ptr[n];
          ft = x_top_ptr->freq[0];
          if (ft > freq)
        break;      // The remaining transitions will have higher thresholds
          if (freq < x_top_ptr->freq[x_top_ptr->np - 1])
        {
        /*Need the appropriate density at this point. */

          nconf = x_top_ptr->nlev;
          density = den_config (xplasma, nconf);
          if (density > DENSITY_PHOT_MIN)
            {
              kappa_tot += x =
            sigma_phot_topbase (x_top_ptr, freq) * density;
/* I believe most of next steps are totally diagnsitic; it is possible if 
statement could be deleted entirely 060802 -- ksl */
              if (geo.ioniz_or_extract) // 57h -- ksl -- 060715
            {   // Calculate during ionization cycles only
              frac_tot += z = x * (freq - ft) / freq;
              nion = config[nconf].nion;
              if (nion > 3)
                {
                  frac_z += z;
                }
              frac_ion[nion] += z;
              kappa_ion[nion] += x;
            }

            }

        }
```

**Comment by jhmatthews** [comment]
I believe the fix here should just be analogous to what is done in resonate.c. We take a photon at the original position, and create a copy at +ds. We then calculate the doppler shifts at these two velocities/positions and take the average of that to compare. We are therefore shifting the photon frequency from the global rest frame into the local rest frame. 

This section of code before the above, instead of just setting freq to p->freq:

``` C
  /* Create phot, a photon at  +ds */
  stuff_phot (p, &phot);
  move_phot (&phot, ds);
  vwind_xyz (&phot, v_outer);
  v2 = dot (phot.lmn, v_outer);

  vwind_xyz (p, v_inner);
  v1 = dot (p->lmn, v_inner);

  freq_inner = p->freq * (1. - v1 / C);
  freq_outer = phot.freq * (1. - v2 / C);

  freq = 0.5 * (freq_inner * freq_outer)

```

**Comment by kslong** [comment]
This seems reasonable as long as the difference between v1 and v2 is small, but in principle you could do better if the frequency is near the ionization edges, where you might not have any absorption part way through the cell and the absorption beyond that.  It's probably not important, but I think I would put in a Warning or Error os some kind if v1 and v2 differ by more than some number, maybe 300 km/s.  This would create a potential problem at and edge.  If this occurs a lot, then we could then check if the two frequencies at the edges span the photoionization edge.
K

**Comment by jhmatthews** [comment]
Proposed improvement here as follows:

``` C
if (ft > freq_min && ft < freq_max)
          {
            frac_path = (ft - freq_min) / (freq_max - freq_min);  // then the shifting off the photon causes it to cross an edge. Find out where between fmin and fmax the edge would be    
            n_cross++;
          }

          else if (ft > freq_max)
        break;      // The remaining transitions will have higher thresholds

          else if (ft < freq_min)
        frac_path = 1.0;        // then all frequency along ds are above edge
```

This checks if an edge is crossed along the path element ds. If it is, then we multiply the opacity by a factor frac_path, where I've just assume that frequency varies linearly between the points:

``` C
kappa_tot += x = sigma_phot_topbase (x_top_ptr, freq) * density * frac_path ;
```

I've also changed the frequency we actually use to work out the heating so that it is always above the threshold (i.e. if we cross an edge over ds it becomes the average of the highest frequency along ds and the threshold frequency, because the inner edge frequency would be less than thrshold, ft, if we crossed an edge) 

``` C
frac_tot += z = x * (freq - ft) / freq;
```

**Comment by jhmatthews** [comment]
![test_edge_fix](https://cloud.githubusercontent.com/assets/3007249/2852353/29caa720-d133-11e3-8183-8541b8383888.png)

Fix 1 is with the first fix, where you take the average frequency. Fix 2 is the new fix where you check if an edge has been crossed.

This fix makes no discernible difference to a spectrum around the Balmer jump. I've checked and this is because an edge only gets crosses every ~1000 or so times a photon travels ds.

the Pull request contains the fix. I've now merged and the issue can be closed.

If we ever want to do better than this, then we would need to step along the path element ds and work out the cross section at each point and relevant opacity. I see no real need to do this, and it may be a time sink. This fix is probably already more than is really needed.
